src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php

```diff
@@ -0,0 +1,358 @@
+            }
+        } else {
+            $directory = dirname($this->file);
+
+            if (!is_dir($directory) && !@mkdir($directory, 0777 & (~$this->umask), true)) {
+                throw new InvalidArgumentException(sprintf('Cache directory does not exist and cannot be created: %s', $directory));
+            }
+
+            if (!is_writable($directory)) {
+                throw new InvalidArgumentException(sprintf('Cache directory is not writable: %s', $directory));
+            }
+        }
+
+        $dump = <<<'EOF'
+<?php
+
+// This file has been auto-generated by the Symfony Cache Component.
+
+return array(
+
+
+EOF;
+
+        foreach ($values as $key => $value) {
+            CacheItem::validateKey($key);
+
+            if (null === $value || is_object($value)) {
+                try {
+                    $value = serialize($value);
+                } catch (\Exception $e) {
+                    throw new InvalidArgumentException(sprintf('Cache key "%s has non-serializable %s value', $key, get_class($value)), 0, $e);
+                }
+            } elseif (is_array($value)) {
+                try {
+                    $serialized = serialize($value);
+                    $unserialized = unserialize($serialized);
+                } catch (\Exception $e) {
+                    throw new InvalidArgumentException(sprintf('Cache key "%s has non-serializable array value', $key), 0, $e);
+                }
+                // Store arrays serialized if they contain any objects or references
+                if ($unserialized !== $value || (false !== strpos($serialized, ';R:') && preg_match('/;R:[1-9]/', $serialized))) {
+                    $value = $serialized;
+                }
+            } elseif (is_string($value)) {
+                // Serialize strings if they could be confused with serialized objects or arrays
+                if ('N;' === $value || (isset($value[2]) && ':' === $value[1])) {
+                    $value = serialize($value);
+                }
+            } elseif (!is_scalar($value)) {
+                throw new InvalidArgumentException(sprintf('Cache key "%s has non-serializable %s value', $key, gettype($value)));
+            }
+
+            $dump .= var_export($key, true).' => '.var_export($value, true).",\n";
+        }
+
+        $dump .= "\n);\n";
+        $dump = str_replace("' . \"\\0\" . '", "\0", $dump);
+
+        $tmpFile = uniqid($this->file);
+
+        file_put_contents($tmpFile, $dump);
+        @chmod($tmpFile, 0666 & (~$this->umask));
+        unset($serialized, $unserialized, $value, $dump);
+
+        @rename($tmpFile, $this->file);
+
+        $this->values = (include $this->file) ?: array();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItem($key)
+    {
+        if (null === $this->values) {
+            $this->initialize();
+        }
+
+        if (!isset($this->values[$key])) {
```

Comment:
let's validate before that $key is a string to prevent issues with magic array index casting (same below)